File: src/VB_GenericControls/DataGrid/ValidationDataGridObjects/PropertyRule.vb
Imports System.ComponentModel

Public Class PropertyRule
    Implements INotifyPropertyChanged

#Region "Construction"

    ''' <summary>
    ''' Construct new property rule.
    ''' </summary>
    ''' <param name="rule">Rule as a function that returns a boolean.</param>
    ''' <param name="message">The error message to display if function returns True.</param>
    Public Sub New(rule As Func(Of Boolean), message As String)
        _rules.Add(New Rule(rule, message))
    End Sub

#End Region

#Region "Members"

    ' This has to implement notify property changed to alert the UI to change color state.
    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
    Private ReadOnly _rules As New List(Of Rule)
    Private _hasError As Boolean = False
    Private _errorMessage As String = String.Empty

    ''' <summary>
    ''' Determines whether the property has an erro. 
    ''' </summary>
    Public Property HasError() As Boolean
        Get
            Return _hasError
        End Get
        Set
            If _hasError <> Value Then
                _hasError = Value
                NotifyPropertyChanged(NameOf(HasError))
            End If
        End Set
    End Property

    ''' <summary>
    ''' Returns the error message as string.
    ''' </summary>
    Public Property ErrorMessage() As String
        Get
            Return _errorMessage
        End Get
        Set
            If _errorMessage <> Value Then
                _errorMessage = Value
                NotifyPropertyChanged(NameOf(ErrorMessage))
            End If
        End Set
    End Property

    Public ReadOnly Property Rules As List(Of Rule)
        Get
            Return _rules
        End Get
    End Property

    ''' <summary>
    ''' Class for the property rule. Each rule has a function and an error message.
    ''' </summary>
    Public Class Rule
        Public ReadOnly Expression As Func(Of Boolean)
        Public ReadOnly Message As String
        Public HasError As Boolean
        Friend Sub New(expression As Func(Of Boolean), message As String)
            Me.Expression = expression
            Me.Message = message
        End Sub
    End Class

#End Region

#Region "Methods"

    ''' <summary>
    ''' Raise property changed event.
    ''' </summary>
    ''' <param name="propertyName">Name of the property than changed.</param>
    Protected Sub NotifyPropertyChanged(propertyName As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub

    ''' <summary>
    ''' Add rule to the proprty.
    ''' </summary>
    ''' <param name="rule">Rule as a function that returns a boolean.</param>
    ''' <param name="message">The error message to display if function returns True.</param>
    Friend Sub AddRule(rule As Func(Of Boolean), message As String)
        _rules.Add(New Rule(rule, message))
    End Sub



    ''' <summary>
    ''' Execute the property rules.
    ''' </summary>
    Friend Sub ExecuteRules()
        ErrorMessage = ""
        HasError = False
        Try
            For i As Int32 = 0 To _rules.Count - 1
                If _rules(i).Expression() = True Then
                    '_rules(i).HasError = True
                    HasError = True
                    ErrorMessage += If(i = 0, _rules(i).Message, Environment.NewLine + _rules(i).Message)
                End If
            Next

            'For Each r As Rule In _rules
            '    If r.Expression() = True Then
            '        r.HasError = True
            '        ErrorMessage = ErrorMessage & r.Message & vbLf
            '        HasError = True
            '    End If
            'Next
            'If HasError Then ErrorMessage = ErrorMessage.TrimEnd(New [Char]() {ControlChars.Lf})
        Catch e As Exception
            _errorMessage = e.Message
            HasError = True
        End Try
    End Sub

#End Region

End Class

File: src/GenericControls/DataGrid/ValidationDataGridObjects/PropertyRule.cs
using System.ComponentModel;

namespace GenericControls
{
    /// <summary>
    /// Represents a validation rule for a property, with support for multiple error conditions and messages.
    /// </summary>
    /// <remarks>
    /// <para>
    ///     <b> Authors: </b>
    /// <list type="bullet">
    /// <item><description>
    ///     Woodrow Fields, USACE Risk Management Center, woodrow.l.fields@usace.army.mil
    /// </description></item>
    /// <item><description>
    ///     Haden Smith, USACE Risk Management Center, cole.h.smith@usace.army.mil
    /// </description></item>
    /// </list>
    /// </para>
    /// </remarks>
    public class PropertyRule : INotifyPropertyChanged
    {

        #region Construction

        /// <summary>
        /// Initializes a new instance of the <see cref="PropertyRule"/> class.
        /// </summary>
        /// <param name="rule">A function that returns true when an error condition is met.</param>
        /// <param name="message">The error message to display if the function returns true.</param>
        public PropertyRule(Func<bool> rule, string message)
        {
            _rules.Add(new Rule(rule, message));
        }

        #endregion

        #region Members

        /// <summary>
        /// Occurs when a property value changes. Required for UI binding to update error states.
        /// </summary>
        public event PropertyChangedEventHandler PropertyChanged;
        private readonly List<Rule> _rules = new List<Rule>();
        private bool _hasError = false;
        private string _errorMessage = string.Empty;

        /// <summary>
        /// Gets or sets a value indicating whether the property has a validation error.
        /// </summary>
        public bool HasError
        {
            get
            {
                return _hasError;
            }
            set
            {
                if (_hasError != value)
                {
                    _hasError = value;
                    NotifyPropertyChanged(nameof(HasError));
                }
            }
        }

        /// <summary>
        /// Gets or sets the error message describing the validation failure.
        /// </summary>
        public string ErrorMessage
        {
            get
            {
                return _errorMessage;
            }
            set
            {
                if ((_errorMessage ?? "") != (value ?? ""))
                {
                    _errorMessage = value;
                    NotifyPropertyChanged(nameof(ErrorMessage));
                }
            }
        }

        /// <summary>
        /// Gets the list of validation rules for this property.
        /// </summary>
        public List<Rule> Rules
        {
            get
            {
                return _rules;
            }
        }

        /// <summary>
        /// Class for the property rule. Each rule has a function and an error message.
        /// </summary>
        public class Rule
        {
            /// <summary>
            /// The expression that evaluates to true when an error condition is met.
            /// </summary>
            public readonly Func<bool> Expression;

            /// <summary>
            /// The error message to display when the expression returns true.
            /// </summary>
            public readonly string Message;

            /// <summary>
            /// Indicates whether this rule currently has an error.
            /// </summary>
            public bool HasError;

            /// <summary>
            /// Initializes a validation rule with an expression and error message.
            /// </summary>
            /// <param name="expression">The expression that returns <c>true</c> when the rule fails.</param>
            /// <param name="message">The message to display for the failed rule.</param>
            internal Rule(Func<bool> expression, string message)
            {
                Expression = expression;
                Message = message;
            }
        }

        #endregion

        #region Methods

        /// <summary>
        /// Raises the <see cref="PropertyChanged"/> event.
        /// </summary>
        /// <param name="propertyName">The name of the property that changed.</param>
        protected void NotifyPropertyChanged(string propertyName)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        /// <summary>
        /// Adds a validation rule to the property.
        /// </summary>
        /// <param name="rule">A function that returns true when an error condition is met.</param>
        /// <param name="message">The error message to display if the function returns true.</param>
        internal void AddRule(Func<bool> rule, string message)
        {
            _rules.Add(new Rule(rule, message));
        }



        /// <summary>
        /// Executes all validation rules for this property and updates the error state.
        /// </summary>
        internal void ExecuteRules()
        {
            ErrorMessage = "";
            HasError = false;
            try
            {
                for (int i = 0, loopTo = _rules.Count - 1; i <= loopTo; i++)
                {
                    if (_rules[i].Expression() == true)
                    {
                        HasError = true;
                        ErrorMessage += i == 0 ? _rules[i].Message : Environment.NewLine + _rules[i].Message;
                    }
                }
            }
            catch (Exception e)
            {
                _errorMessage = e.Message;
                HasError = true;
            }
        }

        #endregion

    }
}
